### ISSUE TITLE
Error while processing asynchronous task

### ISSUE THREAD

**Issue by ashishyadav13237** [Eernie/bitbucket-webhooks-plugin#88]
I am seeing the following in my logs, can you please let me know what this is for?

```
2017-12-14 11:17:05,230 ERROR [threadpool:thread-2]  c.a.s.i.c.DefaultTransferableStateManager Error while processing asynchronous task
java.lang.NullPointerException: null
	at nl.topicus.bitbucket.api.PullRequestListener.lambda$onBuildStatusSetEvent$1(PullRequestListener.java:212)
	at com.atlassian.stash.internal.concurrent.DefaultTransferableStateManager$StateTransferringRunnable.run(DefaultTransferableStateManager.java:166)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.lang.Thread.run(Thread.java:748)
	... 1 frame trimmed
```

**Comment by ashishyadav13237** [comment]
I am seeing this as well:

c.a.e.i.LockFreeEventPublisher$Publisher There was an exception thrown trying to dispatch event 'com.atlassian.bitbucket.event.throttle.TicketAcquiredEvent[source=com.atlassian.stash.internal.throttle.SemaphoreThrottleService@247086d6]' from the invoker 'AsyncInvokerBatch{invokers=[SingleParameterMethodListenerInvoker{method=public void com.atlassian.plugins.shortcuts.internal.DefaultKeyboardShortcutManager.handleEvent(java.lang.Object), listener=com.atlassian.plugins.shortcuts.internal.DefaultKeyboardShortcutManager@4ec89e9}, SingleParameterMethodListenerInvoker{method=public void com.atlassian.upm.UpmInformation.handleApplicationStartedEvent(java.lang.Object), listener=com.atlassian.upm.UpmInformation@4061c900}, SingleParameterMethodListenerInvoker{method=public void com.atlassian.upm.mac.HostLicenseUpdatedEventListener.onAnyEvent(java.lang.Object), listener=com.atlassian.upm.mac.HostLicenseUpdatedEventListener@16f82ab0}]}'.
java.util.concurrent.RejectedExecutionException: Task com.atlassian.sal.core.executor.ThreadLocalDelegateRunnable@30f29671 rejected from com.atlassian.stash.internal.event.EventThreadPoolExecutor@55eb0454[Running, pool size = 4, active threads = 4, queued tasks = 4060, completed tasks = 2900555]
        at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2063)
        at com.atlassian.stash.internal.event.EventThreadPoolExecutor.rejectedExecution(EventThreadPoolExecutor.java:88)
        at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:830)
        at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1379)
        at com.atlassian.sal.core.executor.ThreadLocalDelegateExecutorService.execute(ThreadLocalDelegateExecutorService.java:103)
        at com.atlassian.event.internal.AsynchronousAbleEventDispatcher.dispatch(AsynchronousAbleEventDispatcher.java:88)
        at com.atlassian.event.internal.LockFreeEventPublisher$Publisher.dispatch(LockFreeEventPublisher.java:222)
        at com.atlassian.event.internal.LockFreeEventPublisher.publish(LockFreeEventPublisher.java:95)
        at com.atlassian.stash.internal.event.TransactionAwareEventPublisher.internalPublish(TransactionAwareEventPublisher.java:155)
        at com.atlassian.stash.internal.event.TransactionAwareEventPublisher.publish(TransactionAwareEventPublisher.java:99)
        at com.atlassian.stash.internal.throttle.SemaphoreThrottleService.acquireTicket(SemaphoreThrottleService.java:93)
        at com.atlassian.plugin.util.ContextClassLoaderSettingInvocationHandler.invoke(ContextClassLoaderSettingInvocationHandler.java:26)
        at org.eclipse.gemini.blueprint.service.importer.support.internal.aop.ServiceInvoker.doInvoke(ServiceInvoker.java:56)
        at org.eclipse.gemini.blueprint.service.importer.support.internal.aop.ServiceInvoker.invoke(ServiceInvoker.java:60)
        at org.eclipse.gemini.blueprint.service.util.internal.aop.ServiceTCCLInterceptor.invokeUnprivileged(ServiceTCCLInterceptor.java:70)
        at org.eclipse.gemini.blueprint.service.util.internal.aop.ServiceTCCLInterceptor.invoke(ServiceTCCLInterceptor.java:53)
        at org.eclipse.gemini.blueprint.service.importer.support.LocalBundleContextAdvice.invoke(LocalBundleContextAdvice.java:57)
        at com.atlassian.stash.internal.rest.content.ContentResource$2.write(ContentResource.java:317)
        at com.atlassian.bitbucket.rest.util.JsonStreamingOutput.write(JsonStreamingOutput.java:23)
        at com.atlassian.stash.internal.rest.filter.StreamingResponseFilter$ExceptionMappingStreamingOutput.lambda$write$0(StreamingResponseFilter.java:80)
        at com.atlassian.stash.internal.rest.filter.StreamingResponseFilter$AbstractStreamingExceptionMapper.stream(StreamingResponseFilter.java:50)
        at com.atlassian.stash.internal.rest.filter.StreamingResponseFilter$ExceptionMappingStreamingOutput.write(StreamingResponseFilter.java:80)
        at com.atlassian.applinks.core.rest.context.ContextFilter.doFilter(ContextFilter.java:24)
        at com.atlassian.applinks.core.rest.context.ContextFilter.doFilter(ContextFilter.java:24)
        at com.atlassian.applinks.core.rest.context.ContextFilter.doFilter(ContextFilter.java:24)
        at com.atlassian.applinks.core.rest.context.ContextFilter.doFilter(ContextFilter.java:24)
        at com.atlassian.applinks.core.rest.context.ContextFilter.doFilter(ContextFilter.java:24)
        at com.atlassian.analytics.client.filter.UniversalAnalyticsFilter.doFilter(UniversalAnalyticsFilter.java:92)
        at com.atlassian.analytics.client.filter.AbstractHttpFilter.doFilter(AbstractHttpFilter.java:39)
        at com.atlassian.stash.internal.spring.lifecycle.LifecycleJohnsonServletFilterModuleContainerFilter.doFilter(LifecycleJohnsonServletFilterModuleContainerFilter.java:42)
        at com.atlassian.plugin.connect.plugin.auth.scope.ApiScopingFilter.doFilter(ApiScopingFilter.java:81)
        at com.atlassian.stash.internal.spring.lifecycle.LifecycleJohnsonServletFilterModuleContainerFilter.doFilter(LifecycleJohnsonServletFilterModuleContainerFilter.java:42)
        at com.atlassian.stash.internal.spring.security.StashAuthenticationFilter.onSuccess(StashAuthenticationFilter.java:219)
        at com.atlassian.stash.internal.spring.security.StashAuthenticationFilter.doFilter(StashAuthenticationFilter.java:100)
        at com.atlassian.stash.internal.web.auth.BeforeLoginPluginAuthenticationFilter.doInsideSpringSecurityChain(BeforeLoginPluginAuthenticationFilter.java:112)
        at com.atlassian.stash.internal.web.auth.BeforeLoginPluginAuthenticationFilter.doFilter(BeforeLoginPluginAuthenticationFilter.java:75)
        at com.atlassian.security.auth.trustedapps.filter.TrustedApplicationsFilter.doFilter(TrustedApplicationsFilter.java:94)
        at com.atlassian.oauth.serviceprovider.internal.servlet.OAuthFilter.doFilter(OAuthFilter.java:67)
        at com.atlassian.stash.internal.spring.lifecycle.LifecycleJohnsonServletFilterModuleContainerFilter.doFilter(LifecycleJohnsonServletFilterModuleContainerFilter.java:42)
        at com.atlassian.plugin.connect.plugin.auth.oauth2.DefaultSalAuthenticationFilter.doFilter(DefaultSalAuthenticationFilter.java:69)
        at com.atlassian.plugin.connect.plugin.auth.user.ThreeLeggedAuthFilter.doFilter(ThreeLeggedAuthFilter.java:109)
        at com.atlassian.jwt.internal.servlet.JwtAuthFilter.doFilter(JwtAuthFilter.java:32)
        at com.atlassian.analytics.client.filter.DefaultAnalyticsFilter.doFilter(DefaultAnalyticsFilter.java:38)
        at com.atlassian.analytics.client.filter.AbstractHttpFilter.doFilter(AbstractHttpFilter.java:39)
        at com.atlassian.stash.internal.spring.lifecycle.LifecycleJohnsonServletFilterModuleContainerFilter.doFilter(LifecycleJohnsonServletFilterModuleContainerFilter.java:42)
        at com.atlassian.stash.internal.web.auth.BeforeLoginPluginAuthenticationFilter.doBeforeBeforeLoginFilters(BeforeLoginPluginAuthenticationFilter.java:90)
        at com.atlassian.stash.internal.web.auth.BeforeLoginPluginAuthenticationFilter.doFilter(BeforeLoginPluginAuthenticationFilter.java:73)
        at com.atlassian.stash.internal.request.DefaultRequestManager.doAsRequest(DefaultRequestManager.java:89)

**Comment by ashishyadav13237** [comment]
Thanks. I will follow up with Atlassian on that. What is the cause for the first 1?

**Comment by casz** [comment]
which bitbucket version are you using?

**Comment by ashishyadav13237** [comment]
v5.5.1

**Comment by ashishyadav13237** [comment]
and v1.6.2 of this plugin.

**Comment by casz** [comment]
Then I don't understand how line 212 can be null when BuildStatusSetEvent's getBuildStatus is nonull and so is getState call...

https://developer.atlassian.com/static/javadoc/bitbucket-server/5.5.1/build-api/reference/com/atlassian/bitbucket/build/BuildStatusSetEvent.html#getBuildStatus()
https://developer.atlassian.com/static/javadoc/bitbucket-server/5.5.1/build-api/reference/com/atlassian/bitbucket/build/BuildStatus.html#getState()

**Comment by ashishyadav13237** [comment]
How can this be fixed? Not sure why line 212 is null...

**Comment by rman18** [comment]
We are getting this same issue dozens of times an hour...   27 in the last hour.  We are on BB 5.9.1 and 1.6.2 of this addon:


2018-07-10 16:30:00,098 ERROR [threadpool:thread-3]  c.a.s.i.c.DefaultTransferableStateManager Error while processing asynchronous task
java.lang.NullPointerException: null
	at nl.topicus.bitbucket.api.PullRequestListener.lambda$onBuildStatusSetEvent$1(PullRequestListener.java:212)
	at com.atlassian.stash.internal.concurrent.DefaultTransferableStateManager$StateTransferringRunnable.run(DefaultTransferableStateManager.java:166)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.lang.Thread.run(Thread.java:745)
	... 1 frame trimmed

**Comment by casz** [comment]
Do you have any hints @bturner ?

**Comment by casz** [comment]
@ashishyadav13237 @rman18 I suggest you contact Atlassian support as bturner mentioned since this issue is clearly a internal issue.

I guess in the meantime I could add null checks.

**Comment by rman18** [comment]
I have a case open with Atlassian they told me to contact you but the case is still open.

**Comment by casz** [comment]
Clearly the issue is located at BitBucket internals when they say their method should be Nonnull and it is not

**Comment by rman18** [comment]
@bturner I have added your statement to my case with Atlassian.    We are using Oracle 12.

**Comment by bturner** [comment]
I believe I've isolated how this is happening, and I've made some changes in Bitbucket Server (currently up for review internally) to avoid it. I've created [BSERV-10986](https://jira.atlassian.com/browse/BSERV-10986) to track the issue on our end.

Based on tracing the code, I believe the `NullPointerException` happens when multiple `POST`s are made for the same commit ID and build key in a (very) short period of time. Due to limitations in ActiveObjects it's not possible to create a unique constraint on that combination, so it's possible to have multiple rows where the system really only wants one (the latest one). To "fix" that, when it happens, the system first retrieves all existing rows with the same (commit ID, key) combination and deletes them, and then inserts a new row. Build statuses are never actually updated.

When two build statuses are `POST`ed "close" together, though, what can happen is the first build status is created, and an event raised, and then that build status row is deleted from the database before the event handler gets invoked and a second event is raised. Because of how ActiveObjects proxies entities, the actual values for all the properties on the `BuildStatus` object aren't set until they're accessed the first time, which happens when `PullRequestEventListener` calls `getState()` at line 212. At that point, ActiveObjects queries the database to load the values. Since the row has been deleted, though, it doesn't find anything. Rather than throwing a clear exception, it just returns `null`. Hence the `NullPointerException`.

Since it's not possible for me to fix released versions (e.g., 5.9.1 is always going to be capable of failing like this, even if I backport a fix to our 5.9 branch; the fix will only ship in 5.9.2), I'd encourage adding a `null` check in the `PullRequestEventListener` for `event.getBuildStatus().getState()`. The _commit ID_ on the `BuildStatusSetEvent`, and the `BuildStatus` itself, will never be `null`, but _any property_ on the `BuildStatus` is capable of returning `null`, regardless of how it's documented, due to a...let's go with "unfortunate", or perhaps "not well thought out", behavior in ActiveObjects. If `event.getBuildStatus().getState()` returns `null`, I'd just bail out in the event listener. Note that there's no risk of "losing" the event because the `BuildStatus` properties being `null` is indicative that another build status has been created for the same commit+key--which will have its own event.

**Comment by casz** [comment]
Thanks @bturner glad that I pinged you :) so it could be fixed for the future :smile:

**Comment by casz** [comment]
I'll try and get a release out tomorrow. Gonna be busy all day 😅

**Comment by rman18** [comment]
Atlassian sent me a new bitbucket-build jar that has resolved this issue for us.  I'm told the change will be in the newest versions of the App.

**Comment by casz** [comment]
Sweet!

**Comment by casz** [comment]
@ashishyadav13237 @rman18 released as [v1.6.3](https://marketplace.atlassian.com/apps/1215474/post-webhooks-for-bitbucket/version-history#b100600300)
